src/main/java/spark/route/Routes.java

```diff
@@ -0,0 +1,295 @@
+        LOG.debug("Adds route: " + entry);
+        // Adds to end of list
+        routes.add(entry);
+    }
+
+    //can be cached? I don't think so.
+    private Map<String, RouteEntry> getAcceptedMimeTypes(List<RouteEntry> routes) {
+        Map<String, RouteEntry> acceptedTypes = new HashMap<>();
+
+        for (RouteEntry routeEntry : routes) {
+            if (!acceptedTypes.containsKey(routeEntry.acceptedType)) {
+                acceptedTypes.put(routeEntry.acceptedType, routeEntry);
+            }
+        }
+
+        return acceptedTypes;
+    }
+
+    private boolean routeWithGivenAcceptType(String bestMatch) {
+        return !MimeParse.NO_MIME_TYPE.equals(bestMatch);
+    }
+
+    private List<RouteEntry> findTargetsForRequestedRoute(HttpMethod httpMethod, String path) {
+        List<RouteEntry> matchSet = new ArrayList<RouteEntry>();
+        for (RouteEntry entry : routes) {
+            if (entry.matches(httpMethod, path)) {
+                matchSet.add(entry);
+            }
+        }
+        return matchSet;
+    }
+
+    // TODO: I believe this feature has impacted performance. Optimization?
+    private RouteEntry findTargetWithGivenAcceptType(List<RouteEntry> routeMatches, String acceptType) {
+        if (acceptType != null && routeMatches.size() > 0) {
+            Map<String, RouteEntry> acceptedMimeTypes = getAcceptedMimeTypes(routeMatches);
+            String bestMatch = MimeParse.bestMatch(acceptedMimeTypes.keySet(), acceptType);
+
+            if (routeWithGivenAcceptType(bestMatch)) {
+                return acceptedMimeTypes.get(bestMatch);
+            } else {
+                return null;
+            }
+        } else {
+            if (routeMatches.size() > 0) {
+                return routeMatches.get(0);
+            }
+        }
+
+        return null;
+    }
+
+    private boolean removeRoute(HttpMethod httpMethod, String path) {
+        List<RouteEntry> forRemoval = new ArrayList<>();
+
+        for (RouteEntry routeEntry : routes) {
+            HttpMethod httpMethodToMatch = httpMethod;
+
+            if (httpMethod == null) {
+                // Use the routeEntry's HTTP method if none was given, so that only path is used to match.
+                httpMethodToMatch = routeEntry.httpMethod;
+            }
+
+            if (routeEntry.matches(httpMethodToMatch, path)) {
+                LOG.debug("Removing path {}", path, httpMethod == null ? "" : " with HTTP method " + httpMethod);
+
+                forRemoval.add(routeEntry);
+            }
+        }
+
+        return routes.removeAll(forRemoval);
+    }
+
+    /**
+     * Parse and validates a route and adds it
+     *
+     * @param route      the route path
+     * @param acceptType the accept type
+     * @param target     the invocation target
+     */
+    @Deprecated
+    public void add(String route, String acceptType, Object target) {
+        try {
+            int singleQuoteIndex = route.indexOf(SINGLE_QUOTE);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 274 does not find the substring and returns -1, the call to substring() on line 275 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string route.